torch/nn/modules/normalization.py

```diff
@@ -123,18 +123,27 @@ class LayerNorm(Module):
             has learnable per-element affine parameters initialized to ones (for weights)
             and zeros (for biases). Default: ``True``.
 
+    Attributes:
+        weight: the learnable weights of the module of shape
+            :math:`\text{normalized\_shape}` when :attr:`elementwise_affine` is set to ``True``.
+            The values are initialized to 1.
+        bias:   the learnable bias of the module of shape
+                :math:`\text{normalized\_shape}` when :attr:`elementwise_affine` is set to ``True``.
+                The values are initialized to 0.
+
     Shape:
         - Input: :math:`(N, *)`
         - Output: :math:`(N, *)` (same shape as input)
 
     Examples::
 
         >>> # NLP Example
-        >>> batch, sentence_length, embedding = 20, 5, 10
-        >>> embedding = torch.randn(batch, sentence_length, embedding)
-        >>> layer_norm = nn.LayerNorm(embedding)
+        >>> batch, sentence_length, embedding_dim = 20, 5, 10
+        >>> embedding = torch.randn(batch, sentence_length, embedding_dim)
+        >>> layer_norm = nn.LayerNorm(embedding_dim)
         >>> # Activate module
         >>> layer_norm(embedding)
+        >>>
         >>> # Image Example
         >>> N, C, H, W = 20, 5, 10, 10
```

Comment:
Just curious, is this image example realistic? I think almost all the time, LayerNorm is applied over only the embedding dim (and aggregates away the temporal / spatial dims)

Reply:
Yes, it is. As the figure mentioned in [here](https://github.com/pytorch/pytorch/issues/51455#issuecomment-771004522), LayerNorm normalizes over the last three (C, H, W) dimensions in compute vision. In my opinion, LayerNorm should normalize over the last two (seq_len, emb_dim) dimensions in NLP following [the original paper](https://arxiv.org/abs/1607.06450). However, the implementation of LayerNorm in the most popular NLP project (such as [AllenNLP](https://github.com/allenai/allennlp/blob/f877fdc30d18178b88c335fbd92722fb77c42d93/allennlp/modules/layer_norm.py#L6) is applied over only the embedding dim.

Reply:
I think the GroupNorm paper isn't a good reference, because LayerNorm seems to be used differently these days (as you mention, usually in NLP only normalising over embedding_dim).

I think for image example, some realistic example is needed from some established / representative / recent model / paper. Otherwise, it is confusing.


Problem with normalising over height and width as well is that the affine parameters are then tied to fixed height and width and won't support other dimensions. I don't think this is the most common scenario. Or at least this should be discussed.